### PR TITLE
Updated featureUtility InstallUtils to fix NPE issue

### LINKED ISSUE #17093: featureUtility throws NPE with include tag corner case
**Describe the bug**
featureUtility doesn't check variable element with `defaultValue` attribute and throws npe in verbose mode.

In server.xml 
```
<variable name="db_xml" defaultValue="postgresql.xml"/>
<include optional="true" location="${db_xml}"/>
```

./featureUtility isf defaultServer --verbose 
[2021-05-07, 13:22:25:999 EDT] Checking for missing features required by the server ...
[2021-05-07, 13:22:26:040 EDT] file not found for bootstrap.properties under server: defaultServer
<---------------------> 0.00%
java.lang.NullPointerException
	at com.ibm.ws.install.internal.InstallUtils.getFeatures(InstallUtils.java:742)
	at com.ibm.ws.install.internal.Director.getServerFeaturesToInstall(Director.java:475)
	at com.ibm.ws.install.internal.InstallKernelImpl.getServerFeaturesToInstall(InstallKernelImpl.java:266)
	at com.ibm.ws.install.featureUtility.cli.InstallServerAction.installServerFeatures(InstallServerAction.java:174)
	at com.ibm.ws.install.featureUtility.cli.InstallServerAction.execute(InstallServerAction.java:247)
	at com.ibm.ws.install.featureUtility.cli.InstallServerAction.handleTask(InstallServerAction.java:69)
	at com.ibm.ws.install.featureUtility.cli.FeatureAction.handleTask(FeatureAction.java:92)
	at com.ibm.ws.install.featureUtility.FeatureUtilityExecutor.main(FeatureUtilityExecutor.java:58)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.ibm.ws.kernel.boot.cmdline.UtilityMain.internal_main(UtilityMain.java:173)
	at com.ibm.ws.kernel.boot.cmdline.UtilityMain.main(UtilityMain.java:53)

**Steps to Reproduce**

**Expected behavior**
Should parse the variable value from the `defaultValue`. If the variable is set using both `defaultValue` and `value`. `value` takes precedence. 
Also, if the configuration file specified in the optional include element does not exists, it should continue with the installation rather than exiting with a failure. 

**Diagnostic information:**
 - OpenLiberty Version: [e.g. 21.0.0.4]

### PR DESCRIPTION
**Pull request by abdulabedin**
When the variable is not set (not declared as env var) and relative path is used, featureUtility throws NPE.

In this pull request, I changed InstallUtils.java so it checks:

If the include tag is optional, if it is then then the include tag is skipped and a message is logged with the logger level set to fine.

If the include tag is not optional, then it is also skipped but a message is logged with the logger level set to warning.

Fixes https://github.com/OpenLiberty/open-liberty/issues/17093

### CODE REVIEW

**Comment by abdulabedin** [comment]
#build

**Comment by abdulabedin** [comment]
#build

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ public class InstallUtils {
     private static final int LINE_WRAP_COLUMNS = 72;
     public static final String NEWLINE = System.getProperty("line.separator");
     public static final boolean isWindows = System.getProperty("os.name").toLowerCase().startsWith("windows");
-
+    public static boolean opt = true;
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]:
```suggestion
    public static boolean isOptional = true;
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 2]
```diff
@@ -697,18 +697,27 @@ public static Set<String> getFeatures(String serverXml, String xml, Set<String>
                 Node vl = varList.item(j);
                 Element vlElement = (Element) vl;
                 String varName = vlElement.getAttribute("name");
-                String varVal = vlElement.getAttribute("value");
+                String varVal;
+                if (vlElement.getAttribute("value") == "") {
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]:
```suggestion
                if (vlElement.getAttribute("value").isEmpty()) {
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 3]
```diff
@@ -697,18 +697,27 @@ public static Set<String> getFeatures(String serverXml, String xml, Set<String>
...
 
             // parse include tag
             NodeList includeList = element.getElementsByTagName("include");
+
             for (int i = 0; i < includeList.getLength(); i++) {
                 Node il = includeList.item(i);
                 Element ilElement = (Element) il;
                 String location = ilElement.getAttribute("location");
-
+                if (ilElement.getAttribute("optional").equals("false") || ilElement.getAttribute("optional").equals("")) {
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]:
```suggestion
                if (ilElement.getAttribute("optional").equals("false") || ilElement.getAttribute("optional").isEmpty()) {
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 4]
```diff
@@ -734,12 +744,24 @@ public static Set<String> getFeatures(String serverXml, String xml, Set<String>
                 }
             }
         } catch (Exception e) {
-            logger.log(Level.FINE, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_INVALID_SERVER_XML", xml, e.getMessage()));
+            if (opt == false) {
+                logger.log(Level.INFO, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_INVALID_SERVER_XML", xml, e.getMessage()));
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]:
```suggestion
                logger.log(Level.WARNING, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_INVALID_SERVER_XML", xml, e.getMessage()));
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 5]
```diff
@@ -734,12 +744,24 @@ public static Set<String> getFeatures(String serverXml, String xml, Set<String>
...
+
         }
         visitedServerXmls.add(isUrl ? serverXml : realServerXml.toString());
         for (String filepath : newLocations) {
             Path path = Paths.get(filepath);
-            features.addAll(getFeatures(path.toString(), path.getFileName().toString(), visitedServerXmls));
+            if (path.getFileName() != null) {
+                features.addAll(getFeatures(path.toString(), path.getFileName().toString(), visitedServerXmls));
+            } else if (opt == true) {
+                logger.log(Level.INFO, "Optional Path Not Found");
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]: If the include tag is optional, then just show the log when debug trace is enabled
```suggestion
                logger.log(Level.FINE, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_INVALID_SERVER_XML", path, e.getMessage()));
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 6]
```diff
@@ -734,12 +744,24 @@ public static Set<String> getFeatures(String serverXml, String xml, Set<String>
...
+            } else {
+                logger.log(Level.INFO, "Mandatory Path Not Found");
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]:
```suggestion
                logger.log(Level.WARNING, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_INVALID_SERVER_XML", path, e.getMessage()));
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java [hunk 7]
```diff
@@ -734,12 +744,24 @@ public static Set<String> getFeatures(String serverXml, String xml, Set<String>
...
+            } else {
+                logger.log(Level.FINE, Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_INVALID_SERVER_XML", xml, e.getMessage()));
+            }
+
         }
         visitedServerXmls.add(isUrl ? serverXml : realServerXml.toString());
         for (String filepath : newLocations) {
             Path path = Paths.get(filepath);
-            features.addAll(getFeatures(path.toString(), path.getFileName().toString(), visitedServerXmls));
+            if (path.getFileName() != null) {
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]: Are we trying to check if the file exists here? Then we can do something like
`Files.exists(path)`

**Review by jjiwooLim — CHANGES_REQUESTED** [review]
looks good, just minor changes for readability

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java (hunk 7 above)
Comment (abdulabedin) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallUtils.java]: Yes we are checking if the file exists, I will change it to file.exists.

**Comment by abdulabedin** [comment]
#build

**Comment by abdulabedin** [comment]
#libby-bot

**Review by jjiwooLim — APPROVED** [review]
lgtm!

**Comment by abdulabedin** [comment]
#libby-bot

**Pull request merged**